M1FExplained/Chapter10/Exercises/Exercise05.lean

```diff
@@ -26,31 +35,150 @@ lemma part_a (m n a : ℤ) (hmn : Int.gcd m n = 1) (hm : m ∣ a) (hn : n ∣ a)
+  let M : ℤ := Int.gcd m n
+  /- Since m / M and n / M are both integers, our choice of a has both factors m and n. However m * n > a,
+  so cannot possibly divide it. -/
+  use (m * n / M)
+  apply And.intro
+  -- First prove that m ∣ a.
+  ·rw [show m * n / M = m * (n / M) by {
+    exact Int.mul_ediv_assoc m (show M ∣ n by exact Int.gcd_dvd_right m n)
+  }]
+   have : m ∣ m := by use 1; norm_num
+   exact dvd_mul_of_dvd_left this (n / M)
+  apply And.intro
+  -- Next, prove n ∣ a.
+  ·rw [show m * n / M = n * (m / M) by {
+    rw [←mul_comm n m]
+    exact Int.mul_ediv_assoc n (show M ∣ m by exact Int.gcd_dvd_left m n)
+  }]
+   have : n ∣ n := by use 1; norm_num
+   exact dvd_mul_of_dvd_left this (m / M)
+  have : M ∣ m := by exact Int.gcd_dvd_left m n
+  have hMmn : M ∣ m * n := by {
+    match this with
+    |⟨k, hk⟩ => 
+    use k * n
+    rw [hk]
+    rw [mul_assoc]}
+  -- Finally, we prove that (m * n) cannot divide a, because it is larger.
+  ·have h1' : 0 < abs (m * n / M) := by {
+    rw [abs_pos]
+    intro h
+    have := Int.eq_zero_of_ediv_eq_zero hMmn h
+    rw [mul_eq_zero] at this
+    tauto
+  }
+   have h1M : 1 < abs M := by {
+    apply by_contradiction
+    intro h
+    push_neg at h
+    have := @Int.abs_le_one_iff M
+    rw [this] at h
+    apply Or.elim h
+    ·intro hM0
+     simp at hM0
+     rw [Int.gcd_eq_zero_iff] at hM0
+     tauto
+    ·intro
+     | Or.inl hone => simp at hone; contradiction
+     | Or.inr hmone => 
+     have : 0 < M := by {
+      rw [@Int.ofNat_pos]
+      exact Int.gcd_pos_of_ne_zero_left n hm
+     }
+     contradiction
+   }
+   -- Apply both helper_1 and helper_2 to finish the proof.
+   have h2' : abs (m * n / M)  < abs (m * n) := by exact helper_1 m n M h1M hMmn hm hn
+   exact helper_2 (m * n / M) (m * n) h1' h2'
+
+-- part c
+
+/-
+Show that if gcd(x, m) = 1 and gcd(y, m) = 1, then gcd(x * y, m) = 1
+-/
 
 lemma part_c (x y m : ℤ) (hx : Int.gcd x m = 1) (hy : Int.gcd y m = 1) : Int.gcd (x * y) m = 1 := by 
   let a := Int.gcdA x m
   let b := Int.gcdB x m
   let c := Int.gcdA y m
   let d := Int.gcdB y m
+  -- Apply Bézout's lemma to x, m and y, m respectivley.
   have hxm : 1 = x * a + m * b := by rw [←Int.gcd_eq_gcd_ab x m, hx]; norm_cast
   have hym : 1 = y * c + m * d := by rw [←Int.gcd_eq_gcd_ab y m, hy]; norm_cast
-  have h   : 1 = x * y * a * c + (x * a * d + b * y * c + m * b * d) * m := by
+  have h   : 1 = a * c * x * y  + (x * a * d + b * y * c + m * b * d) * m := by
     calc 
       1 = 1 * 1 := by norm_num
       _ = (x * a + m * b) * (y * c + m * d) := by nth_rewrite 1 [hxm]; rw [hym]
       _ = x * a * y * c + x * a * m * d + m * b * y * c + m * b * m * d := by ring
-      _ = x * y * a * c + (x * a * d + b * y * c + m * b * d) * m := by ring
-  /- Need lemma saying that if ∃ s, t : a * s * b * t = 1, then gcd(a, b) = 1, this is Q4C -/
-  sorry
-
-
+      _ = a * c * x * y  + (x * a * d + b * y * c + m * b * d) * m := by ring
```

Comment:
You don't need two consecutive `by ring`s in the calculation; just delete the one above this one.